python/OmniFoldTTbar.py

```diff
@@ -263,23 +264,40 @@ def _get_input_arrays(self, preprocess=True):
         # truth level
         arr_gen = self.handle_sig.get_arrays(self.varnames_truth, valid_only=False)
 
-        if preprocess:
-            logger.info("Preprocess feature arrays")
+        # preprocess input arrays
 
-            # estimate the order of magnitude
-            logger.debug("Divide each variable by its order of magnitude")
-            # use only the valid events
-            xmean_reco = np.mean(np.abs(self.handle_obs[self.varnames_reco]), axis=0)
-            xoom_reco = 10**(np.log10(xmean_reco).astype(int))
-            arr_data /= xoom_reco
-            arr_sim /= xoom_reco
+        p = preprocessor.get()
 
-            xmean_truth = np.mean(np.abs(self.handle_sig[self.varnames_truth]), axis=0)
-            xoom_truth = 10**(np.log10(xmean_truth).astype(int))
-            arr_gen /= xoom_truth
+        arr_data = p.preprocess(self.varnames_reco, arr_data)
+        # TODO: a truth known option?
+        
+        arr_sim = p.preprocess(self.varnames_reco, arr_sim)
+        arr_gen = p.preprocess(self.varnames_truth, arr_gen)
```

Comment:
I'm also wondering for the mapping of a feature to something else e..g phi to cos(phi) would it be better to make this a part of data handler? This kind of mapping could be done for sim and data separately when extracting the feature arrays.
I guess the other kind of the preprocessing that involves all feature arrays e.g. standardize to mean of zero and variance of one have to be done when we put them together. This can probably be done outside of `_get_input_arrays` in `run`, where we have `X_data`, `X_sim`, `X_gen`. What do you think?

Reply:
The single feature array preprocessing (angle_to_sin, for example) can technically be done at the data handler level. The multiple array preprocessing like standardization needs to be done at at least the _get_input_arrays level if not later.

The raised issue with dummy values is right. I think it does make sense to move preprocessing into run. We get the `valid_only=False` data at the beginning of `run` with `X_data, X_sim, X_gen = self._get_input_arrays()`. Then it gets the pass cut flags. So I think we can preprocess the valid part of the feature arrays instead the entirety of it.

Reply:
Sounds good!
Yes, the second type of preprocessing like standardization should really be applied on the feature arrays that are actually used in the training (and validation). This step could essentially be a preprocessing layer in the network. We could even put this inside omnifold.py and apply on X_step1 and X_step2?

Reply:
> We could even put this inside omnifold.py and apply on X_step1 and X_step2?
That's an excellent idea! Then we don't need to deal with pairing and pass cut issues. I think I'll move the 2nd preprocessing into omnifold.

Reply:
lol apparently we are thinking the same thing

Reply:
One disadvantage of doing this though is it is not convenient to plot the distributions of feature arrays separately for data and sim after preprocessing. But I guess it's ok.

Reply:
That's true. We don't have that currently so I think it is ok.

Reply:
We do currently plot the arrays for training after preprocessing in `run` [here](https://github.com/zctao/omnifoldTop/blob/master/python/OmniFoldTTbar.py#L371). We could do these plots using `X_step1[Y_step1==1]` etc.